include/spinlock.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _SPINLOCK_H
+#define _SPINLOCK_H
+
+#include <atomic.h>
+
+struct k_spinlock_key {
+	int key;
+};
+
+typedef struct k_spinlock_key k_spinlock_key_t;
+
+struct k_spinlock {
+#ifdef CONFIG_SMP
+	atomic_t locked;
+#ifdef CONFIG_DEBUG
+	int saved_key;
+#endif
+#endif
+};
+
+static inline k_spinlock_key_t k_spin_lock(struct k_spinlock *l)
+{
+	k_spinlock_key_t k;
+
+	/* Note that we need to use the underlying arch-specific lock
+	 * implementation.  The "irq_lock()" API in SMP context is
+	 * actually a wrapper for a global spinlock!
+	 */
+	k.key = _arch_irq_lock();
```

Comment:
Just for my own understanding: this locks interrupts just on the current CPU or all of them in this context?
If this is global I might be being thick but I am unsure how this works.

Reply:
Only on the current CPU, because it's in a critical section while the spinlock is held.  The other CPUs can do their own thing until/unless they want to contend for the same lock.